spec/predicates/runtime-trace.md

```diff
@@ -0,0 +1,2712 @@
+-   Minor version changes are always backwards compatible and “monotonic.” Such
+    changes do not update the `predicateType`.
+-   Producers MAY add extension fields using field names that are URIs.
+-   Fields marked _optional_ MAY be unset or null, and should be treated
+    equivalently. Both are equivalent to empty for object or array values.
+
+## Fields
+
+`monitor` _object_, _required_
+
+Identifies the specific monitor instance used to trace the runtime.
+
+`monitor.id` _string (URI)_, _required_
+
+URI indicating the monitor’s identity.
+
+`monitor.configSource` ResourceDescriptor, _optional_
+
+Effectively a pointer to the monitor's configuration.
+
+`monitor.tracePolicy` _object_, _optional_
+
+Indicates the trace policy used for the monitoring event that generated the
+runtime trace. The trace policy's format is dependent on the monitor used, and
+so it must be parsed after identifying the monitor using `monitor.id`.
+
+FIXME: should this be optional?
+
+`monitoredProcess` _object_, _required_
+
+Identifies the process being monitored.
+
+`monitoredProcess.hostID` _string (URI)_, _required_
+
+URI indicating the process host’s identity. Ex: GitLab Runner ID
+
+`monitoredProcess.type` _string (TypeURI)_, _required_
+
+URI indicating the type of process performed. Ex: when monitoring a build, this
+field records the build type.
+
+`monitoredProcess.event` _string_, _required_
+
+String identifying the specific job or task associated with the attestation.
+
+`monitorLog` _object_, _required_
+
+Record of events that were traced by the monitor. One of `process`, `network`,
+and `fileAccess` must be present.
+
+`monitorLog.process` _list_, _optional_
+
+`monitorLog.network` _list_, _optional_
+
+`monitorLog.fileAccess` _list_, _optional_
+
+Record of files accessed during the build process. A complete list of
+_materials_ can be derived from this information. Each entry in this list is
+expected to record the path of the file and one or more digests of the file.
+This field is a list rather than a key value map because a single file may be
+used multiple times during the build process. Further, some files that are
+accessed may _change_ during the build process, and so, different entries may
+have different digests.
+
+While this predicate can be used to log file accesses, the actual technique used
+to capture the file access event has some implications. If a synchronous
+monitor, for example one that uses `ptrace` to trace the file access system
+calls, is used, then the build process can be paused while the file's digest is
+calculated and stored. However, asynchronous monitors such as those using eBPF
+cannot pause the build process before the file is actually used. Therefore, they
+cannot make as strong guarantees about the digests of the files accessed.
+Verifiers using runtime trace attestations for file accesses must be careful
+about what guarantees they are actually getting based on how the build process
+was monitored.
```

Comment:
IMO this is overly detailed and tailored for an assumed use case, while other elements of the monitorLog are left undefined.

Reply:
Correct, we were worried about using file hashes recorded here to validate certain guarantees and wanted to be explicit that in some cases, the files may have changed between being accessed by the process and their hashes being recorded. Certainly erring on the side of caution, I'm open to hearing alternatives to this. Perhaps this could be a general specification-wide disclaimer as well about how attestations are captured and used. Until then, I'd prefer to leave this in here.